Option for excluding noise parameters in design matrix.

abhisrkckl opened the pull request:
Related to issue #1310. This PR does not fully address this issue as it is still under discussion. However, it does
(1) Raise a NotImplementedError if the user tries to compute the design matrix with a noise parameter unfrozen with a clearer error message. Earlier, it used to raise an AttributeError complaining about missing derivative functions.
(2) Provide an option to exclude noise parameters from the design matrix.

I have made the default behavior of the TimingModel.designmatrix() function the same as before to avoid breaking any existing code.

aarchiba commented on tests/test_designmatrix_noisepars.py:
```diff
@@ -0,0 +1,46 @@
+@pytest.fixture()
+def model_and_toas():
+    datadir = pinttestdata.datadir
+    parfile = os.path.join(datadir, "J1713+0747_NANOGrav_11yv0_short.gls.par")
```
Does this par file have the noise parameters unfrozen? What makes it special? Any reason not to use a simpler one with a set of TOAs that is much faster to load?

abhisrkckl replied:
I just picked one which had EFACs, EQUADs, and ECORRs.

aarchiba replied:
Can you create a really simple one in-line? This test is likely to take unnecessarily long if you use this very large data set. It would also be best if it were obvious what is needed from the par file.

abhisrkckl replied:
OK. I will create a smaller dataset for this.

aarchiba commented on tests/test_designmatrix_noisepars.py:
```diff
@@ -0,0 +1,46 @@
+    return mb.get_model_and_toas(parfile, timfile)
+
+
+def _test_designmatrix(model, toas):
```
Better to directly express the test conditions in place rather than call this utility function - it makes it hard to tell from the test code what the expected behaviour is.

abhisrkckl replied:
OK.. I'll remove the utility function.

aarchiba commented on src/pint/models/timing_model.py:
```diff
@@ -1744,6 +1756,8 @@ def designmatrix(self, toas, acc_delay=None, incfrozen=False, incoffset=True):
             Whether to include frozen parameters in the design matrix
         incoffset : bool
             Whether to include the constant offset in the design matrix
+        incnoise : bool
```
Will it ever be a good idea to include noise parameters in the design matrix? Is there a reason to create this non-functional parameter?

abhisrkckl replied:
I did it this way to ensure that the default behavior of PINT remains unchanged so that nothing else breaks.

aarchiba replied:
I think this parameter should be removed, and noise parameters not included. Any code that expects them to be included is currently broken, since none of them have correct derivatives.

abhisrkckl replied:
I'll remove the incnoise parameter.

aarchiba commented on tests/test_designmatrix_noisepars.py:
```diff
@@ -0,0 +1,46 @@
+    M = model.designmatrix(toas, incnoise=False)
+
+
+def test_designmatrix_noise_exception_falsepositive(model_and_toas):
```
This title is unclear. Perhaps in this situation no noise parameter is free and you want to make sure the design matrix can be created? Can the title say that?

abhisrkckl replied:
I have changed the function name so that it is clearer.

aarchiba commented on src/pint/models/timing_model.py:
```diff
@@ -552,6 +552,16 @@ def get_params_dict(self, which="free", kind="quantity"):
                 raise ValueError(f"Unknown kind {kind!r}")
         return c
 
+    def get_params_of_component_type(self, component_type):
```
This function needs documentation.

abhisrkckl replied:
Added dosctring.

abhisrkckl commented on tests/test_designmatrix_noisepars.py:
```diff
@@ -0,0 +1,46 @@
+    try:
+        # This should raise a NotImplementedError rather than an AttributeError.
+        M = model.designmatrix(toas)
+    except NotImplementedError:
```
OK.. I'll change that.

dlakaplan commented:
@aarchiba : should this be merged?

aarchiba commented on tests/test_designmatrix_noisepars.py:
```diff
@@ -0,0 +1,62 @@
+
+@pytest.fixture()
+def model_and_toas():
+    datadir = pinttestdata.datadir
```
Better to just do `from pinttestdata import datadir`, above.

aarchiba replied:
You can also use `datadir / "file.tim"` instead of `os.path.join`.

aarchiba commented on tests/test_designmatrix_noisepars.py:
```diff
@@ -0,0 +1,62 @@
+    return mb.get_model_and_toas(parfile, timfile)
+
+
+def test_designmatrix_noise_exception_no_free_noise_param(model_and_toas):
```
This name seems to suggest that it expects an exception?

abhisrkckl replied:
Yes.. That's confusing. I'll rename that.

aarchiba commented on tests/test_designmatrix_noisepars.py:
```diff
@@ -0,0 +1,62 @@
+    # This should work.
+    M = model.designmatrix(toas, incnoise=False)
+
+    model.EFAC1.frozen = True
```
The model and TOAs are created anew for each test, so no need to worry about resetting it.

abhisrkckl replied:
I was not sure about this. I thought the same objects were used for all tests.

aarchiba replied:
The `pytest` documentation on fixtures has some detail on this, but the idea is to ensure that tests cannot contaminate each other's results by mangling the fixture objects. This does mean that fixture objects are re-created frequently, so it is important to ensure that they are as light-weight as possible.

abhisrkckl commented:
I have made the following changes:
1. Created a new dataset with a simpler par file and less number of TOAs
removed the 'incnoise' parameter

abhisrkckl commented:
I have made the following changes:
1. Created a new dataset with a simpler par file and less number of TOAs
2. removed the 'incnoise' parameter
3. Simplified the unit test

aarchiba commented on tests/test_designmatrix_noisepars.py:
```diff
@@ -0,0 +1,33 @@
+    model, toas = model_and_toas
+
+    # This should work.
+    M = model.designmatrix(toas)
```
Are there any properties M should have that you can `assert` here? Or if you just want the function call to not raise an exception, maybe say so in the function name? (Also you don't have to assign the return value to anything if you never use it.)

abhisrkckl replied:
I have added new assertions to test for the shape of the design matrix.

aarchiba commented on tests/test_designmatrix_noisepars.py:
```diff
@@ -0,0 +1,33 @@
+    M = model.designmatrix(toas)
+
+
+def test_designmatrix_free_noise_params(model_and_toas):
```
Please use `pytest.mark.parametrize` to automatically turn this into three tests, one for each value of `noiseparam`. This will ensure separate pass/failure for each of the three, and also simplify the test and avoid the need to try to reset the model at the end of each one.

abhisrkckl commented:
The failed CI test is due to a numpy/astropy version issue.

paulray commented:
For some reason the version specification in oldestdeps is `1.17.*` but the version being installed is `1.23.3`. Not sure why this is happening, but it is causing all of our oldestdeps CI runs to fail.
